MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace OpenOrchestra\MediaAdmin\Reference\Strategies;
+
+use OpenOrchestra\Media\Repository\MediaRepositoryInterface;
+use OpenOrchestra\BBcodeBundle\Parser\BBcodeParser;
+use OpenOrchestra\Media\BBcode\AbstractMediaCodeDefinition;
+
+/**
+ * Class AbstractReferenceStrategy
+ */
+abstract class AbstractMediaReferenceStrategy
+{
+    protected $mediaRepository;
+    protected $bbcodeParser;
+
+    /**
+     * MediaRepositoryInterface $mediaRepository
+     * BBcodeParser             $bbcodeParser
+     */
+    public function __construct(MediaRepositoryInterface $mediaRepository, BBcodeParser $bbcodeParser)
+    {
+        $this->mediaRepository = $mediaRepository;
+        $this->bbcodeParser = $bbcodeParser;
+    }
+
+    /**
+     * Recursively extract media ids from elements (bloc, attribute, collection attribute, etc ...)
+     *
+     * @param mixed $element
+     * @param array $references
+     *
+     * @return array
+     */
+    protected function extractMediasFromElement($element, array $references = array())
+    {
+        if ($this->isMediaAttribute($element)) {
+            $references[] = $element['id'];
+        } elseif (is_string($element) && $this->hasMediaBBcode($element)) {
+            $references = $this->extractMediaFromBBCode($element, $references);
+        } elseif (is_array($element)) {
+            foreach ($element as $item) {
+                $references = $this->extractMediasFromElement($item, $references);
+            }
+        }
+
+        return $references;
+    }
+
+    /**
+     * @param string $str
+     * @param array  $references
+     *
+     * @return array
+     */
+    protected function extractMediaFromBBCode($str, array $references)
```

Comment:
This method don't really need $references as parameter. Merge should be done in the calling method.